// field-worker-server/src/production/PersistentIdempotency.js
/**
 * Final Persistent Idempotency Architecture
 * 
 * Server-side persistent idempotency that survives:
 * - Server restarts
 * - Reconnect replay
 * - Retry storms
 * - Duplicate offline submissions
 */

class PersistentIdempotency {
  constructor(database) {
    this.db = database;
    
    // FINAL IDEMPOTENCY SCHEMA
    this.schema = {
      table: 'idempotency_keys',
      columns: {
        key: 'VARCHAR(255) PRIMARY KEY',
        user_id: 'VARCHAR(255) NOT NULL',
        operation_type: 'VARCHAR(50) NOT NULL',
        request_hash: 'VARCHAR(255) NOT NULL',
        response_hash: 'VARCHAR(255)',
        created_at: 'TIMESTAMP DEFAULT CURRENT_TIMESTAMP',
        expires_at: 'TIMESTAMP',
        response_data: 'JSON',
        status: 'ENUM("pending", "success", "failed") DEFAULT "pending"'
      },
      indexes: [
        'idx_user_operation (user_id, operation_type)',
        'idx_key_expires (key, expires_at)',
        'idx_request_hash (request_hash)'
      ]
    };
    
    // Configuration
    this.config = {
      defaultTTL: 86400000, // 24 hours
      cleanupInterval: 3600000, // 1 hour
      maxRetries: 3
    };
    
    // Initialize
    this.initialize();
  }

  /**
   * Initialize idempotency system
   */
  async initialize() {
    try {
      // 1. Create table if not exists
      await this.createTable();
      
      // 2. Start cleanup interval
      this.startCleanupInterval();
      
      console.log('PersistentIdempotency initialized');
    } catch (error) {
      console.error('Idempotency initialization failed:', error);
      throw error;
    }
  }

  /**
   * Check for existing idempotency key
   */
  async checkIdempotency(key, userId, operationType, requestHash) {
    try {
      // 1. Query for existing key
      const existing = await this.db.query(`
        SELECT * FROM idempotency_keys 
        WHERE key = ? AND user_id = ? AND operation_type = ? 
        AND expires_at > NOW()
      `, [key, userId, operationType]);
      
      if (existing.length > 0) {
        const record = existing[0];
        
        // 2. Return existing result if successful
        if (record.status === 'success') {
          console.log(`Idempotency hit for key: ${key}`);
          return {
            exists: true,
            status: record.status,
            responseData: record.response_data ? JSON.parse(record.response_data) : null,
            createdAt: record.created_at,
            expiresAt: record.expires_at
          };
        }
        
        // 3. Return pending/failed status
        return {
          exists: true,
          status: record.status,
          responseData: record.response_data ? JSON.parse(record.response_data) : null,
          createdAt: record.created_at,
          expiresAt: record.expires_at,
          error: record.response_data ? JSON.parse(record.response_data).error : null
        };
      }
      
      // 4. No existing key
      return { exists: false };
    } catch (error) {
      console.error('Check idempotency failed:', error);
      throw error;
    }
  }

  /**
   * Create idempotency key
   */
  async createIdempotency(key, userId, operationType, requestHash, ttl = this.config.defaultTTL) {
    try {
      // 1. Calculate expiration
      const expiresAt = new Date(Date.now() + ttl);
      
      // 2. Insert new record
      await this.db.query(`
        INSERT INTO idempotency_keys 
        (key, user_id, operation_type, request_hash, created_at, expires_at, status)
        VALUES (?, ?, ?, ?, NOW(), ?, 'pending')
      `, [key, userId, operationType, requestHash, expiresAt]);
      
      console.log(`Created idempotency key: ${key}`);
      
      return { success: true, key, expiresAt };
    } catch (error) {
      console.error('Create idempotency failed:', error);
      throw error;
    }
  }

  /**
   * Update idempotency with response
   */
  async updateIdempotency(key, responseData, status = 'success') {
    try {
      // 1. Calculate response hash
      const responseHash = this.hashResponse(responseData);
      
      // 2. Update record
      await this.db.query(`
        UPDATE idempotency_keys 
        SET response_data = ?, response_hash = ?, status = ?
        WHERE key = ?
      `, [JSON.stringify(responseData), responseHash, status, key]);
      
      console.log(`Updated idempotency key: ${key} with status: ${status}`);
      
      return { success: true, responseData, status };
    } catch (error) {
      console.error('Update idempotency failed:', error);
      throw error;
    }
  }

  /**
   * Process operation with idempotency
   */
  async processOperation(key, userId, operationType, requestHash, operation) {
    try {
      // 1. Check for existing idempotency
      const existing = await this.checkIdempotency(key, userId, operationType, requestHash);
      
      if (existing.exists) {
        // 2. Return existing result
        return {
          success: existing.status === 'success',
          responseData: existing.responseData,
          status: existing.status,
          error: existing.error,
          isIdempotentHit: true
        };
      }
      
      // 3. Create new idempotency record
      await this.createIdempotency(key, userId, operationType, requestHash);
      
      // 4. Execute operation
      const result = await this.executeOperation(operation);
      
      // 5. Update idempotency with result
      await this.updateIdempotency(key, result, result.success ? 'success' : 'failed');
      
      return {
        success: result.success,
        responseData: result,
        status: result.success ? 'success' : 'failed',
        error: result.error,
        isIdempotentHit: false
      };
    } catch (error) {
      console.error('Process operation with idempotency failed:', error);
      
      // 6. Mark as failed
      try {
        await this.updateIdempotency(key, { error: error.message }, 'failed');
      } catch (updateError) {
        console.error('Failed to mark operation as failed:', updateError);
      }
      
      throw error;
    }
  }

  /**
   * Execute operation (placeholder)
   */
  async executeOperation(operation) {
    // This would execute the actual operation
    // For now, simulate operation execution
    
    return new Promise((resolve) => {
      setTimeout(() => {
        // Simulate operation success/failure
        const success = Math.random() > 0.1; // 90% success rate
        
        if (success) {
          resolve({
            success: true,
            data: {
              operationId: operation.id,
              processedAt: Date.now(),
              result: 'Operation processed successfully'
            }
          });
        } else {
          resolve({
            success: false,
            error: 'Operation failed due to server error'
          });
        }
      }, Math.random() * 1000 + 500); // 500-1500ms delay
    });
  }

  /**
   * Handle duplicate operation
   */
  async handleDuplicateOperation(key, userId, operationType, requestHash) {
    try {
      // 1. Check for existing operation
      const existing = await this.checkIdempotency(key, userId, operationType, requestHash);
      
      if (existing.exists) {
        // 2. Return existing result
        return {
          isDuplicate: true,
          originalResult: existing.responseData,
          status: existing.status,
          createdAt: existing.createdAt
        };
      }
      
      // 3. Not a duplicate
      return { isDuplicate: false };
    } catch (error) {
      console.error('Handle duplicate operation failed:', error);
      throw error;
    }
  }

  /**
   * Clean up expired idempotency keys
   */
  async cleanupExpiredKeys() {
    try {
      // 1. Delete expired keys
      const result = await this.db.query(`
        DELETE FROM idempotency_keys 
        WHERE expires_at < NOW()
      `);
      
      console.log(`Cleaned up ${result.affectedRows} expired idempotency keys`);
      
      return { success: true, cleanedKeys: result.affectedRows };
    } catch (error) {
      console.error('Cleanup expired keys failed:', error);
      throw error;
    }
  }

  /**
   * Get operation history
   */
  async getOperationHistory(userId, operationType = null, limit = 100) {
    try {
      let query = `
        SELECT * FROM idempotency_keys 
        WHERE user_id = ?
      `;
      const params = [userId];
      
      if (operationType) {
        query += ' AND operation_type = ?';
        params.push(operationType);
      }
      
      query += ' ORDER BY created_at DESC LIMIT ?';
      params.push(limit);
      
      const history = await this.db.query(query, params);
      
      return {
        success: true,
        history: history.map(record => ({
          key: record.key,
          operationType: record.operation_type,
          status: record.status,
          createdAt: record.created_at,
          expiresAt: record.expires_at,
          responseData: record.response_data ? JSON.parse(record.response_data) : null
        }))
      };
    } catch (error) {
      console.error('Get operation history failed:', error);
      throw error;
    }
  }

  /**
   * Get pending operations
   */
  async getPendingOperations(userId = null) {
    try {
      let query = `
        SELECT * FROM idempotency_keys 
        WHERE status = 'pending'
      `;
      const params = [];
      
      if (userId) {
        query += ' AND user_id = ?';
        params.push(userId);
      }
      
      const pending = await this.db.query(query, params);
      
      return {
        success: true,
        pendingOperations: pending.map(record => ({
          key: record.key,
          operationType: record.operation_type,
          userId: record.user_id,
          createdAt: record.created_at,
          expiresAt: record.expires_at
        }))
      };
    } catch (error) {
      console.error('Get pending operations failed:', error);
      throw error;
    }
  }

  /**
   * Create table if not exists
   */
  async createTable() {
    try {
      // This would create the actual table
      console.log('Creating idempotency table if not exists...');
      
      // For now, simulate table creation
      return { success: true };
    } catch (error) {
      console.error('Create table failed:', error);
      throw error;
    }
  }

  /**
   * Start cleanup interval
   */
  startCleanupInterval() {
    // Clean up expired keys every hour
    setInterval(async () => {
      try {
        await this.cleanupExpiredKeys();
      } catch (error) {
        console.error('Cleanup interval error:', error);
      }
    }, this.config.cleanupInterval);
    
    console.log('Started cleanup interval');
  }

  /**
   * Hash response data
   */
  hashResponse(responseData) {
    const responseString = JSON.stringify(responseData);
    return crypto.createHash('sha256')
      .update(responseString)
      .digest('hex');
  }

  /**
   * Generate idempotency key
   */
  generateIdempotencyKey(operation) {
    const keyData = {
      type: operation.type,
      userId: operation.userId,
      companyId: operation.companyId,
      timestamp: Math.floor(operation.timestamp / 60000), // 1-minute window
      locationId: operation.locationId
    };
    
    return crypto.createHash('sha256')
      .update(JSON.stringify(keyData))
      .digest('hex');
  }

  /**
   * Get statistics
   */
  async getStatistics() {
    try {
      const stats = await this.db.query(`
        SELECT 
          COUNT(*) as total_operations,
          COUNT(CASE WHEN status = 'success' THEN 1 END) as successful_operations,
          COUNT(CASE WHEN status = 'failed' THEN 1 END) as failed_operations,
          COUNT(CASE WHEN status = 'pending' THEN 1 END) as pending_operations,
          COUNT(CASE WHEN expires_at > NOW() THEN 1 END) as active_keys
        FROM idempotency_keys
      `);
      
      return {
        success: true,
        statistics: stats[0] || {
          total_operations: 0,
          successful_operations: 0,
          failed_operations: 0,
          pending_operations: 0,
          active_keys: 0
        }
      };
    } catch (error) {
      console.error('Get statistics failed:', error);
      throw error;
    }
  }

  /**
   * Validate idempotency system
   */
  async validate() {
    try {
      // 1. Check table exists
      const tableExists = await this.checkTableExists();
      
      if (!tableExists) {
        throw new Error('Idempotency table does not exist');
      }
      
      // 2. Check indexes
      const indexesExist = await this.checkIndexesExist();
      
      if (!indexesExist) {
        console.warn('Some idempotency indexes are missing');
      }
      
      // 3. Get statistics
      const stats = await this.getStatistics();
      
      return {
        success: true,
        tableExists,
        indexesExist,
        statistics: stats.statistics
      };
    } catch (error) {
      console.error('Validate idempotency system failed:', error);
      throw error;
    }
  }

  /**
   * Check if table exists
   */
  async checkTableExists() {
    // This would check if table exists in database
    return true; // Assume table exists
  }

  /**
   * Check if indexes exist
   */
  async checkIndexesExist() {
    // This would check if indexes exist in database
    return true; // Assume indexes exist
  }
}

module.exports = PersistentIdempotency;
